Add new ThinBox type for 1 stack pointer wide heap allocated trait objects

yaahc opened the pull request:
**Zulip Thread**: https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/ThinBox

Based on https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs

Tracking Issue: https://github.com/rust-lang/rust/issues/92791

Usage Trial: https://github.com/yaahc/pgx/pull/1/files

## TODO

- [x] make sure to test with #[repr(align(1024))] structs etc

yaahc commented:
> Did you intentionally move `box.rs` to `box/mod.rs`? It feels like unnecessary churn.

yea, I didn't want to put all of `ThinBox` in the same module but I can undo that if others prefer.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3945168875) [(plain)](https://github.com/rust-lang/rust/commit/59072239a2a3448c15c10c5ccd287caa0051cf48/checks/3945168875/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
* 627 error codes
* highest error code: E0785
tidy error: /checkout/src/test/ui/box/thin_align.rs:3: TODO is deprecated; use FIXME
Found 0 error codes with no tests
Done!
* 353 features
some tidy checks failed
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:10
```

</details>

yaahc commented:
cc @matthieu-m since you wrote https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs

bors commented:
:umbrella: The latest upstream changes (presumably #90067) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

yaahc commented:
> Do feel free to reuse as much or as little as you wish from https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs, I open-sourced it under a permissive license so that others could build upon it.

Perfect. I figured as much but I always prefer asking before copying other people's code.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4768699649?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/184f0567752af0780841ae58a39fdbc9ad6c1377/checks/4768699649/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling alloc v0.0.0 (/checkout/library/alloc)
   Compiling cfg-if v0.1.10
   Compiling adler v0.2.3
   Compiling rustc-demangle v0.1.21
error[E0658]: use of unstable library feature 'ptr_metadata'
 --> library/alloc/src/boxed/thin.rs:4:23
  |
4 | use core::ptr::{self, Pointee};
  |
  = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
  = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
  = help: add `#![feature(ptr_metadata)]` to the crate attributes to enable
error[E0658]: use of unstable library feature 'ptr_metadata'
error[E0658]: use of unstable library feature 'ptr_metadata'
  --> library/alloc/src/boxed/thin.rs:14:11
   |
14 |     meta: <Dyn as Pointee>::Metadata,
   |
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = help: add `#![feature(ptr_metadata)]` to the crate attributes to enable
error[E0658]: use of unstable library feature 'ptr_metadata'
error[E0658]: use of unstable library feature 'ptr_metadata'
  --> library/alloc/src/boxed/thin.rs:52:19
52 |         let ptr = ptr::from_raw_parts(value, metadata);
   |                   ^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = help: add `#![feature(ptr_metadata)]` to the crate attributes to enable
error[E0658]: use of unstable library feature 'ptr_metadata'
error[E0658]: use of unstable library feature 'ptr_metadata'
  --> library/alloc/src/boxed/thin.rs:34:51
   |
34 |         let (_, metadata) = (u_ref as *const Dyn).to_raw_parts();
   |
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = note: see issue #81513 <https://github.com/rust-lang/rust/issues/81513> for more information
   = help: add `#![feature(ptr_metadata)]` to the crate attributes to enable
For more information about this error, try `rustc --explain E0658`.
error: could not compile `alloc` due to 4 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

yaahc commented:
@matthieu-m I updated the PR to utilize your design. I swapped all of the `NonNull` usage to be working with direct raw pointers since that seemed to be the primary diff between your rfc2580 and the current unstable APIs. I think it's all working right now, or at least it is passing the alignment test I added which the previous version failed on, but I'm relatively inexperienced at writing unsafe code like this in rust so if you have time to do a quick review I'd greatly appreciate it.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4781095809?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5f7bc99ebe0d81edcadb0b33d67b2ef8880e9765/checks/4781095809/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
..........................................iii....................................................... 12500/12547
...............................................
failures:

---- [ui] ui/box/thin_new.rs stdout ----
error: test compilation failed although it shouldn't!
status: exit status: 1
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/box/thin_new.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/box/thin_new/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/box/thin_new/auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
error[E0308]: mismatched types
  --> /checkout/src/test/ui/box/thin_new.rs:8:34
   |
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
LL |     let _a: ThinBox<dyn Error> = ThinBox::new(Foo);
   |             ------------------   ^^^^^^^^^^^^^^^^^ expected trait object `dyn std::error::Error`, found struct `Foo`
   |             expected due to this
   |
   |
   = note: expected struct `ThinBox<dyn std::error::Error>`
              found struct `ThinBox<Foo>`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.

---
test result: FAILED. 12427 passed; 1 failed; 119 ignored; 0 measured; 0 filtered out; finished in 145.64s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "ui" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:25
```

</details>

compiler-errors commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,165 @@
+///
+/// A thin pointer, regardless of T.
+#[unstable(feature = "thin_box", issue = "92791")]
+pub struct ThinBox<T: ?Sized + Pointee> {
+    ptr: WithHeader<T::Metadata>,
```
`T` always implements `Pointee`, so it might be clearer to remove `Pointee` bounds (here and elsewhere) and just write uses of `T::Metadata` like `<T as Pointee>::Metadata`:

```suggestion
pub struct ThinBox<T: ?Sized> {
    ptr: WithHeader<<T as Pointee>::Metadata>,
```

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+    fn deref_mut(&mut self) -> &mut T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts::<T>(value as *const (), metadata) as *mut T;
```
```suggestion
        let pointer = ptr::from_raw_parts_mut::<T>(value as *mut (), metadata);
```

matthieu-m commented:
> @matthieu-m I updated the PR to utilize your design. I swapped all of the `NonNull` usage to be working with direct raw pointers since that seemed to be the primary diff between your rfc2580 and the current unstable APIs. I think it's all working right now, or at least it is passing the alignment test I added which the previous version failed on, but I'm relatively inexperienced at writing unsafe code like this in rust so if you have time to do a quick review I'd greatly appreciate it.

There's a very good reason to use `NonNull` though: it allows compiler optimizations for the `Option<ThinBox<T>>` case.

That is:

- With raw pointer: `size_of::<Option<ThinBox<u8>>() == 16`.
- With `NonNull`: `size_of::<Option<ThinBox<u8>>() == 8`.

When I played around with `Pointee`, `NonNull` implemented all the necessary APIs:

- [`NonNull::from_raw_parts`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.from_raw_parts)
- [`NonNull::to_raw_parts`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.to_raw_parts)

And I'd suggest adding a check that `size_of::<Option<ThinBox<T>>() == size_of::<*const ()>()` for a normal type, a slice, and a trait to ensure this doesn't regress (either as test, or debug assert).

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+    {
+        let dyn_ref: &Dyn = &value;
+        let (_, meta) = (dyn_ref as *const Dyn).to_raw_parts();
+        let ptr = WithHeader::new(meta, value).expect("No allocation failure");
```
Ibid: I think `alloc::handle_alloc_error` would be more appropriate here.

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+    pub fn new(value: T) -> Self {
+        let dyn_ref: &T = &value;
+        let (_, meta) = (dyn_ref as *const T).to_raw_parts();
+        let ptr = WithHeader::new(meta, value).expect("No allocation failure");
```
I think `alloc::handle_alloc_error` would be more appropriate here.

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+    }
+}
+
+struct WithHeader<H>(*mut u8, PhantomData<H>);
```
Using `NonNull` here would add a niche so `Option<ThinBox<T>>` would fit in a `usize`.

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+        let aligned_header_size = Self::aligned_header_size(layout.align());
+
+        unsafe {
+            let ptr = NonNull::new(alloc::alloc(layout))?;
```
It would be interesting to avoid calling `alloc::alloc` if both `T` and `H` are ZST.

yaahc replied:
I'm not exactly sure how to do that, though that does sound fun. I think I'd like to leave that for a followup PR, though I will add some tests right now to make sure our ZST behavior isn't completely broken.

yaahc replied:
Actually though, what tests can I even run to ensure thinbox is working when the T and H are both ZSTs? My understanding is that would mean that the vtable is empty so it would have to be a trait object for something like `trait Useless {}`. what can I even do with that type?

a1phyr replied:
Nevermind, using `alloc::System` does this for us if we add a generic allocator parameter to `ThinBox`, which will be done as a follow-up, so no need to bother about that small optimization now.

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+
+    //  Safety:
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: *mut T) {
```
This should definitely take a `&mut`

ben0x539 replied:
It's probably fine because `&self` points at the `NonNull<u8>` while `value` points at (the neighborhood of) the pointee, so it's not actually overlapping, right?

eddyb replied:
> Also, using the pointer here may violate an aliasing rule (not sure, but this signature seems dubious).

cc @RalfJung but my understanding was that between `&mut T` and `*mut T`, the latter was more flexible, and the former is the one that can cause more aliasing issues

thomcc replied:
Yes, it seems a lot less dubious to use a pointer here.

yaahc replied:
I swapped it back based on the consensus, marking this as resolved now.

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,198 @@
+#[unstable(feature = "thin_box", issue = "92791")]
+pub struct ThinBox<T: ?Sized> {
+    ptr: WithHeader<<T as Pointee>::Metadata>,
+    _marker: PhantomData<fn(T) -> T>,
```
I assume this `PhantomData` is there for invariance, but I think (not sure though) that using `<T as Pointee>::Metadata` is enough here. Additionally, there should be a `PhantomData<T>` for dropck.

yaahc replied:
> but I think (not sure though) that using `<T as Pointee>::Metadata` is enough here

Can you explain how for me? I believe you I just don't understand how.

a1phyr replied:
Even if `T` can coerce to `U`, the compiler cannot guarantee that `<T as Pointee>::Metadata` can coerce to `<U as Pointee>::Metadata`, which makes the struct invariant.

[Playground example](https://play.rust-lang.org/?gist=b91fa44920e7dfcc47535d52be7c6f15)

yaahc replied:
Okay, that makes sense, I went ahead and updated the PhantomData to T so I think this should be resolved, thank you!

a1phyr reviewed (commented):
This a great idea, thanks for bringing this to `std` !

compiler-errors commented:
@matthieu-m, @yaahc - should we use actually [`Unique<T>`](https://docs.rs/ptr/latest/ptr/struct.Unique.html) instead of just `NonNull<T>`, since this type acts semantically like a `Box` and we get more things for free (e.g. `Send`/`Sync`)?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4793441514?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/7858d8e846888868cfeee92104c03f7ca616456d/checks/4793441514/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

4 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (6 moderate, 2 high)
To address issues that do not require attention, run:
  npm audit fix


Some issues need review, and may require choosing

Run `npm audit` for details.
npm notice 
npm notice 
npm notice New major version of npm available! 7.21.1 -> 8.3.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.3.0>
npm notice Run `npm install -g npm@8.3.0` to update!
Removing intermediate container 0a703f409a34
 ---> 4a5db7006d47
Step 6/12 : RUN npm install eslint@8.6.0 -g
 ---> Running in 744ab166b914
---
Successfully built da33e0eab003
Successfully tagged rust-ci:latest
Built container sha256:da33e0eab003b374501ebf7532945fc9094cb7ef2067316247e97c7277f9f866
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rand v0.7.3
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
error: function is never used: `want_thin`
   |
13 | fn want_thin() {
   |    ^^^^^^^^^
   |
   |
   = note: `-D dead-code` implied by `-D warnings`
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:54
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4793586868?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/0803b5e0579fdc95c0c0325dccdcbaf21a544314/checks/4793586868/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

4 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (6 moderate, 2 high)
To address issues that do not require attention, run:
  npm audit fix


Some issues need review, and may require choosing

Run `npm audit` for details.
npm notice 
npm notice 
npm notice New major version of npm available! 7.21.1 -> 8.3.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.3.0>
npm notice Run `npm install -g npm@8.3.0` to update!
Removing intermediate container 6fd656ee3ffd
 ---> d1bc08c1db49
Step 6/12 : RUN npm install eslint@8.6.0 -g
 ---> Running in 55157546aaa9
---
Successfully built 560945eebcbf
Successfully tagged rust-ci:latest
Built container sha256:560945eebcbf9235df33d20dafc11c6511d1bcf34b179e6beedc4061d9e5e7c3
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rand v0.7.3
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
    Checking alloc v0.0.0 (/checkout/library/alloc)
error: function is never used: `want_thin`
   |
13 | fn want_thin() {
   |    ^^^^^^^^^
   |
   |
   = note: `-D dead-code` implied by `-D warnings`
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:47
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4793788874?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/63f2e45f0af7f5e6c16c6eb7f1a0342f82298487/checks/4793788874/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

4 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (6 moderate, 2 high)
To address issues that do not require attention, run:
  npm audit fix


Some issues need review, and may require choosing

Run `npm audit` for details.
npm notice 
npm notice 
npm notice New major version of npm available! 7.21.1 -> 8.3.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.3.0>
npm notice Run `npm install -g npm@8.3.0` to update!
Removing intermediate container 52872c12cf89
 ---> e7d7ea92a674
Step 6/12 : RUN npm install eslint@8.6.0 -g
 ---> Running in 0a58270836d3
---
Successfully built 39179b3a7b15
Successfully tagged rust-ci:latest
Built container sha256:39179b3a7b15f180ba90458e0f8e89f51a839b2d0859f69708f81f0db7e05a2a
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rand v0.7.3
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
    Checking alloc v0.0.0 (/checkout/library/alloc)
error: function is never used: `want_thin`
   |
13 | fn want_thin() {
   |    ^^^^^^^^^
   |
   |
   = note: `-D dead-code` implied by `-D warnings`
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:57
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4793896136?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/763cf7341e4942deaa671efb282e6e452ca61af2/checks/4793896136/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

4 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (6 moderate, 2 high)
To address issues that do not require attention, run:
  npm audit fix


Some issues need review, and may require choosing

Run `npm audit` for details.
npm notice 
npm notice 
npm notice New major version of npm available! 7.21.1 -> 8.3.0
npm notice Changelog: <https://github.com/npm/cli/releases/tag/v8.3.0>
npm notice Run `npm install -g npm@8.3.0` to update!
Removing intermediate container 905545cca522
 ---> 98088cbffef1
Step 6/12 : RUN npm install eslint@8.6.0 -g
 ---> Running in b02be9c6f14a
---
Successfully built d2bd730cda9b
Successfully tagged rust-ci:latest
Built container sha256:d2bd730cda9ba4e3112143c09f7f33b5e85a7e44b739e254a91c69c56fe5d1a3
Uploading finished image to https://ci-caches.rust-lang.org/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef
upload failed: - to s3://rust-lang-ci-sccache2/docker/9567406ca397026e6d08b99ab2979859301d063de191c2500807ab84c19e827d07128981858917062997c4fe33329efa040c86f471172809b1d85076deabb2ef Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
    Checking rand v0.7.3
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking std v0.0.0 (/checkout/library/std)
    Checking core v0.0.0 (/checkout/library/core)
error: function is never used: `want_thin`
   |
10 | fn want_thin() {
   |    ^^^^^^^^^
   |
   |
   = note: `-D dead-code` implied by `-D warnings`
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
Build completed unsuccessfully in 0:01:34
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4794814963?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b8f4fe97fa4461bd23c892ef9a39a54a3e85d89b/checks/4794814963/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 5a34a5416bd9
Successfully tagged rust-ci:latest
Built container sha256:5a34a5416bd9a7f9235abac39a64789e2e7e56773c3cd5f97310937252f56366
Uploading finished image to https://ci-caches.rust-lang.org/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0
upload failed: - to s3://rust-lang-ci-sccache2/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0 Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-llvm-12]
---
status: exit status: 2
command: "make"
stdout:
------------------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling  --edition=2018 --crate-type=rlib ../../../../library/alloc/src/lib.rs --cfg feature=\"external_crate\" --cfg no_global_oom_handling
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
error[E0425]: cannot find function `handle_alloc_error` in module `alloc`
   --> ../../../../library/alloc/src/boxed/thin.rs:156:24
156 |                 alloc::handle_alloc_error(layout);
    |                        ^^^^^^^^^^^^^^^^^^ not found in `alloc`

warning: unused import: `from_fn`
warning: unused import: `from_fn`
  --> ../../../../library/alloc/src/string.rs:51:18
   |
51 | use core::iter::{from_fn, FusedIterator};
   |
   = note: `#[warn(unused_imports)]` on by default

error: aborting due to previous error; 1 warning emitted
error: aborting due to previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0425`.
make: *** [Makefile:4: all] Error 1
------------------------------------------




failures:
    [run-make] run-make-fulldeps/alloc-no-oom-handling

test result: FAILED. 209 passed; 1 failed; 19 ignored; 0 measured; 0 filtered out; finished in 31.83s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--rust-demangler-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" "--src-base" "/checkout/src/test/run-make-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-12/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:35:02
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4795404227?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/40d4ac37860aeb6352ed321ec9b143932be25686/checks/4795404227/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built 51af5793d6f7
Successfully tagged rust-ci:latest
Built container sha256:51af5793d6f7ac931dba0c407b3fe0ae0e4496a4327c64f60fab50bcf5d9685e
Uploading finished image to https://ci-caches.rust-lang.org/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0
upload failed: - to s3://rust-lang-ci-sccache2/docker/93608764ca974412130ba4a304c50b0ea4a89a9b5853d488b55e530021cbeebcb5549b17247c6674f12706902ca4687aae74f60088fb03044e2d4026dd4d59e0 Unable to locate credentials
[CI_JOB_NAME=x86_64-gnu-llvm-12]
---
status: exit status: 2
command: "make"
stdout:
------------------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps/alloc-no-oom-handling/alloc-no-oom-handling  --edition=2018 --crate-type=rlib ../../../../library/alloc/src/lib.rs --cfg feature=\"external_crate\" --cfg no_global_oom_handling
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
error[E0425]: cannot find function `handle_alloc_error` in module `alloc`
   --> ../../../../library/alloc/src/boxed/thin.rs:158:24
158 |                 alloc::handle_alloc_error(layout);
    |                        ^^^^^^^^^^^^^^^^^^ not found in `alloc`

warning: unused import: `Unsize`
warning: unused import: `Unsize`
 --> ../../../../library/alloc/src/boxed/thin.rs:7:33
  |
7 | use core::marker::{PhantomData, Unsize};
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused import: `from_fn`
warning: unused import: `from_fn`
  --> ../../../../library/alloc/src/string.rs:51:18
   |
51 | use core::iter::{from_fn, FusedIterator};

error: aborting due to previous error; 2 warnings emitted

For more information about this error, try `rustc --explain E0425`.
For more information about this error, try `rustc --explain E0425`.
make: *** [Makefile:4: all] Error 1
------------------------------------------




failures:
    [run-make] run-make-fulldeps/alloc-no-oom-handling

test result: FAILED. 209 passed; 1 failed; 19 ignored; 0 measured; 0 filtered out; finished in 36.21s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--rust-demangler-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" "--src-base" "/checkout/src/test/run-make-fulldeps" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make-fulldeps" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make-fulldeps" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-12/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:36:10
```

</details>

a1phyr commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,211 @@
+
+            //  Safety:
+            //  -   The size is at least `aligned_header_size`.
+            let ptr = ptr.offset(aligned_header_size as isize) as *mut _;
```
This offset calculation should happen after the `NonNull` check

yaahc commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+struct WithHeader<H>(NonNull<u8>, PhantomData<H>);
```
From https://github.com/rust-lang/rust/pull/90066#issuecomment-1011277371

Should this be `Unique` instead of `NonNull`?

compiler-errors replied:
Regarding my comment, I don't think we gain anything by making this `Unique`. I forgot that we're not storing a `NonNull<T>` but a `NonNull<u8>`, and we are already storing `PhantomData<T>` in the `ThinBox` itself :)

eddyb commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,219 @@
+
+    //  Safety:
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: &mut T) {
```
Isn't this risky? I remember there being issues in the past with an (argument) reference's backing storage being deallocated mid-way through a function. Something about the (argument) reference being guaranteed alive at the start vs "throughout". Might need someone else to explain it though.

Maybe running this in miri, with stacked borrows enabled, might be enough to detect an issue? Not sure though, the raw pointers might mask it partly.

Simplest change AFAICT (if this is a problem) would be to make it `value: *mut T` instead, and turn it into a reference only for `{size,align}_of_val`.

yaahc replied:
I updated it from `value: *mut T` because of [this comment](https://github.com/rust-lang/rust/pull/90066#discussion_r783283147) by @a1phyr. I'm not really sure who is right or what to do so I'm gonna also cc @matthieu-m who originally wrote the code and did use `*mut T` so they can weigh in as well since this is one of the last unresolved comments atm.

cuviper replied:
I think #55005 is an example of what @eddyb means.

yaahc replied:
Okay, this should be resolved now.

RalfJung replied:
Yeah using a raw ptr seems better.

thomcc commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    fn aligned_header_size(alloc_align: usize) -> usize {
+        cmp::max(Self::header_size(), alloc_align)
```
This looks a bit suspicious, esp in conjunction with how alloc size is computed above.

I think this will work for currently existing types, but will fail for things where with metadata larger than the value (or header) alignment, but not aligned to value align.

For example, metadata type as `[u8; 17]`, and value type as `u64` (I know this is a bit silly, but we should probably not leave this bug until later). Then the alloc_align will be 8 (for u64), `align_of::<H>()` will be 1, and we'll end up storing the u64 at an unaligned address. (Sorry if this isn't explained well)

So, I think this should be aligning/rounding up here, rather than computing the max. I also like https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.extend for this kind of thing, although perhaps it's annoying to integrate here without a performance overhead from the `Result`. You could use it to check the values in `alloc_layout` in debug builds, though.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4824467977?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/bd7027ca6dd704f8ce0ec81743d1ecc35a06a6c4/checks/4824467977/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/alloc/src/boxed/thin.rs at line 2:
 // https://github.com/matthieu-m/rfc2580/blob/b58d1d3cba0d4b5e859d3617ea2d0943aaa31329/examples/thin.rs
 // by matthieu-m
 use crate::alloc::{self, Layout, LayoutError};
-use core::hint::unreachable_unchecked;
 use core::fmt::{self, Debug, Display, Formatter};
+use core::hint::unreachable_unchecked;
 use core::marker::{PhantomData, Unsize};
 use core::mem;
 use core::ops::{Deref, DerefMut};
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/benches/btree/set.rs" "/checkout/library/alloc/src/boxed/thin.rs" "/checkout/library/alloc/src/rc/tests.rs" "/checkout/library/alloc/src/task.rs" "/checkout/library/alloc/src/boxed.rs" "/checkout/library/alloc/benches/vec.rs" "/checkout/library/alloc/benches/string.rs" "/checkout/library/alloc/src/string.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92598) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

yaahc commented:
second attempt!

r? rust-lang/libs

bors commented:
:umbrella: The latest upstream changes (presumably #95552) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

joshtriplett commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,216 @@
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: *mut T) {
+        unsafe {
+            let Ok((layout, value_offset)) = Self::alloc_layout(Layout::for_value_raw(value)) else {
+                // SAFETY: Layout must have been computable if we're in drop
+                unreachable_unchecked()
+            };
```
Isn't this `unwrap_unchecked`?

```suggestion
            // SAFETY: Layout must have been computable if we're in drop
            let (layout, value_offset) = Self::alloc_layout(Layout::for_value_raw(value)).unwrap_unchecked();
```

joshtriplett commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,216 @@
+    }
+}
+
+struct WithHeader<H>(NonNull<u8>, PhantomData<H>);
```
```suggestion
/// A pointer to type-erased data, guaranteed to have a header `H` before the pointed-to location.
struct WithHeader<H>(NonNull<u8>, PhantomData<H>);
```

joshtriplett commented on library/alloc/src/boxed/thin.rs:
```diff
@@ -0,0 +1,216 @@
+        self.0.as_ptr()
+    }
+
+    fn header_size() -> usize {
```
Could this be `const`?

```suggestion
    const fn header_size() -> usize {
```

(Not entirely sure if this is within our current const capabilities.)

joshtriplett commented on library/alloc/src/lib.rs:
```diff
@@ -152,6 +153,7 @@
 #![feature(fundamental)]
 #![cfg_attr(not(test), feature(generator_trait))]
 #![feature(lang_items)]
+#![feature(let_else)]
```
:tada:

joshtriplett commented:
Posted a few comments, and un-resolved one other comment that didn't get addressed.

LGTM otherwise; r=me with those addressed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5875842555?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/193fbb58434f89e67f29aedd8110b96c9de07c28/checks/5875842555/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/alloc/src/boxed/thin.rs at line 180:
     unsafe fn drop<T: ?Sized>(&self, value: *mut T) {
         unsafe {
             // SAFETY: Layout must have been computable if we're in drop
-            let (layout, value_offset) = Self::alloc_layout(Layout::for_value_raw(value)).unwrap_unchecked();
+            let (layout, value_offset) =
+                Self::alloc_layout(Layout::for_value_raw(value)).unwrap_unchecked();
             ptr::drop_in_place::<T>(value);
             ptr::drop_in_place::<T>(value);
             // We only drop the value because the Pointee trait requires that the metadata is copy
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/alloc/src/macros.rs" "/checkout/library/alloc/benches/vec_deque.rs" "/checkout/library/alloc/src/tests.rs" "/checkout/library/alloc/benches/vec.rs" "/checkout/library/alloc/src/boxed/thin.rs" "/checkout/library/alloc/src/rc.rs" "/checkout/library/alloc/src/borrow.rs" "/checkout/library/alloc/benches/binary_heap.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5876108532?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/051d8a6965cda25f326b31512e9e886f17701714/checks/5876108532/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
37 error: aborting due to 2 previous errors


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coercion/coerce-issue-49593-box-never.nofallback/coerce-issue-49593-box-never.nofallback.stderr
To only update this specific test, also pass `--test-args coercion/coerce-issue-49593-box-never.rs`


error in revision `nofallback`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/coercion/coerce-issue-49593-box-never.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "nofallback" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coercion/coerce-issue-49593-box-never.nofallback" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/coercion/coerce-issue-49593-box-never.nofallback/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0277]: the trait bound `(): std::error::Error` is not satisfied
   |
   |
LL |     /* *mut $0 is coerced to Box<dyn Error> here */ Box::<_ /* ! */>::new(x)
   |                                                     ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::error::Error` is not implemented for `()`
   = help: the following other types implement trait `std::error::Error`:
             !
             &'a T
             AccessError
---

error[E0277]: the trait bound `(): std::error::Error` is not satisfied
  --> /checkout/src/test/ui/coercion/coerce-issue-49593-box-never.rs:22:49
   |
LL |     /* *mut $0 is coerced to *mut Error here */ raw_ptr_box::<_ /* ! */>(x)
   |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::error::Error` is not implemented for `()`
   = help: the following other types implement trait `std::error::Error`:
             !
             &'a T
             AccessError
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #95798) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

yaahc commented:
@bors r=joshtriplett

bors commented:
:pushpin: Commit a87a0d089e793903844ef40851deae48c039f8bb has been approved by `joshtriplett`

<!-- @bors r=joshtriplett a87a0d089e793903844ef40851deae48c039f8bb -->
<!-- homu: {"type":"Approved","sha":"a87a0d089e793903844ef40851deae48c039f8bb","approver":"joshtriplett"} -->
